realm/realm-library/src/androidTest/kotlin/io/realm/UUIDTests.kt

```diff
@@ -0,0 +1,435 @@
+                .schema(UUIDPrimaryKeyRequired::class.java,
+                        UUIDPrimaryKeyNotRequired::class.java,
+                        UUIDAndString::class.java,
+                        UUIDRequiredRealmList::class.java,
+                        UUIDOptionalRealmList::class.java)
+                .build()
+        realm = Realm.getInstance(realmConfiguration)
+    }
+
+    @After
+    fun tearDown() {
+        realm.close()
+    }
+
+    @Test
+    fun copyToAndFromRealm() {
+        val uuid1 = UUID.randomUUID()
+        val uuid2 = UUID.randomUUID()
+        val uuid3 = UUID.randomUUID()
+
+        val value = UUIDPrimaryKeyRequired()
+        value.id = uuid1
+        value.anotherId = uuid2
+        value.name = "Foo"
+
+        // copyToRealm
+        realm.beginTransaction()
+        val obj = realm.copyToRealm(value)
+        realm.commitTransaction()
+        assertEquals(uuid1, obj.id)
+        assertEquals(uuid2, obj.anotherId)
+        assertEquals("Foo", obj.name)
+
+        // copyToRealmOrUpdate
+        value.name = "Bar"
+        value.anotherId = uuid3
+        realm.beginTransaction()
+        realm.copyToRealmOrUpdate(value)
+        realm.commitTransaction()
+
+        // copyFromRealm
+        val copy = realm.copyFromRealm(obj)
+        assertEquals(uuid1, copy.id)
+        assertEquals(uuid3, copy.anotherId)
+        assertEquals("Bar", copy.name)
+    }
+
+    @Test
+    fun insert() {
+        val uuid1 = UUID.randomUUID()
+        val uuid2 = UUID.randomUUID()
+        val uuid3 = UUID.randomUUID()
+
+        val value = UUIDPrimaryKeyRequired()
+        value.id = uuid1
+        value.name = "Foo"
+        value.anotherId = uuid2
+
+        // insert
+        realm.beginTransaction()
+        realm.insert(value)
+        realm.commitTransaction()
+
+        val obj = realm.where<UUIDPrimaryKeyRequired>().findFirst()
+        assertNotNull(obj)
+        assertEquals(uuid1, obj!!.id)
+        assertEquals(uuid2, obj.anotherId)
+        assertEquals("Foo", obj.name)
+
+        // insertOrUpdate
+        realm.beginTransaction()
+        obj.anotherId = uuid3
+        obj.name = "Bar"
+        realm.insertOrUpdate(obj)
+        realm.commitTransaction()
+
+        val all = realm.where<UUIDPrimaryKeyRequired>().findAll()
+        assertEquals(1, all.size)
+        assertEquals(uuid1, all[0]!!.id)
+        assertEquals(uuid3, all[0]!!.anotherId)
+        assertEquals("Bar", all[0]!!.name)
+    }
+
+    @Test
+    fun frozen() {
+        val uuid1 = UUID.randomUUID()
+
+        realm.beginTransaction()
+        val obj = realm.createObject<UUIDPrimaryKeyRequired>(uuid1)
+        obj.name = "foo"
+        realm.commitTransaction()
+
+        val frozen = obj.freeze<UUIDPrimaryKeyRequired>()
+        assertEquals(uuid1, frozen.id)
+        assertEquals("foo", frozen.name)
+    }
+
+    @Test
+    fun requiredPK() {
+        val uuid1 = UUID.randomUUID()
+
+        realm.beginTransaction()
+        try {
```

Comment:
We have `assertFailsWith<RealmException> { ... }` making it a lot easier to use this pattern.